contracts/strategies/dnsVectorStrategy/DnsVectorStrategyInvestmentLib.sol

```diff
@@ -428,4 +427,225 @@ library DnsVectorStrategyInvestmentLib {
+            address(strategyStorage.pangolinRouter),
+            aaveBorrowTokenAmountDesired
+        );
+        (, , uint256 lpTokenAmountChange) = strategyStorage
+            .pangolinRouter
+            .addLiquidity(
+                address(strategyStorage.ammPairDepositToken),
+                address(strategyStorage.aaveBorrowToken),
+                ammPairDepositTokenAmountDesired,
+                aaveBorrowTokenAmountDesired,
+                0,
+                0,
+                address(this),
+                // solhint-disable-next-line not-rely-on-time
+                block.timestamp
+            );
+
+        // 4. Deposit LP token to Pangolin.
+        strategyStorage.pangolinPair.approve(
+            address(strategyStorage.pangolinMiniChef),
+            lpTokenAmountChange
+        );
+        strategyStorage.pangolinMiniChef.deposit(
+            strategyStorage.pangolinPoolId,
+            lpTokenAmountChange,
+            address(this)
+        );
+
+        // 5. Swap remaining `AaveBorrowToken` to `DepositToken`.
+        uint256 aaveBorrowTokenAmountAfter = strategyStorage
+            .aaveBorrowToken
+            .balanceOf(address(this));
+
+        if (aaveBorrowTokenAmountAfter > aaveBorrowTokenAmountBefore) {
+            address[] memory path = new address[](2);
+            path[0] = address(strategyStorage.aaveBorrowToken);
+            path[1] = address(strategyStorage.depositToken);
+
+            uint256 aaveBorrowTokenAmountRemaining = aaveBorrowTokenAmountAfter -
+                    aaveBorrowTokenAmountBefore;
+
+            SwapServiceLib.swapExactTokensForTokens(
+                strategyStorage.swapService,
+                aaveBorrowTokenAmountRemaining,
+                0,
+                path,
+                new uint256[](0)
+            );
+        }
+    }
+
+    function __prepareTokens(
+        uint256 ammPairDepositTokenAmountIn,
+        uint256 aaveBorrowTokenAmountIn
+    )
+        private
+        returns (
+            uint256 ammPairDepositTokenAmountDesired,
+            uint256 aaveBorrowTokenAmountDesired
+        )
+    {
+        DnsVectorStorage storage strategyStorage = DnsVectorStorageLib
+            .getStorage();
+
+        // Calculate which token we have more than we need.
+        (
+            uint256 ammPairDepositTokenReserve,
+            uint256 aaveBorrowTokenReserve
+        ) = DnsVectorStrategyAumLib.getPangolinLpReserve(
+                address(strategyStorage.ammPairDepositToken),
+                address(strategyStorage.aaveBorrowToken)
+            );
+
+        bool swapAmmPairDepositTokenToAaveBorrowToken;
+        bool swapAaveBorrowTokenToAmmPairDepositToken;
+
+        uint256 ammPairDepositTokenDesiredMulByReserve = ammPairDepositTokenAmountIn *
+                aaveBorrowTokenReserve;
+        uint256 aaveBorrowTokenDesiredMulByReserve = aaveBorrowTokenAmountIn *
+            ammPairDepositTokenReserve;
+
+        if (
```

Comment:
I got a different formula and when I substiute into it, it seem to be correct

Reply:
![image](https://user-images.githubusercontent.com/19815616/210226362-8e4e2f1a-80b6-4e98-bfde-52366ed823d3.png)
![image](https://user-images.githubusercontent.com/19815616/210226399-864ba693-d6a0-4e8b-a620-01fd4a2b47c8.png)

Reply:
I think we don't need to contact the price oracle here, what is important is the price implied by liquidity, as rate is used for swapping. I think in that case our formulas will be the same.

Reply:
I'm sorry but I can't recognize your handwriting. Sharing information in a format that is graspable by others is recommended in communication. As you're saying that it's correct, I'll take it as double check.

Using a price of the pool would be optimal if we also swap through the pool. However, as we use other swap service, I think querying oracle a price is better. Also, we don't need to worry about gas cost in this case.

Reply:
I tried my best handwriting but seemed to fail.

It is good that you mention swap service because I think that should be set to either Pangolin or TraderJoe V2, as we don't expect to have much liquidity left in TraderJoeV1. Please confirm with Jasa about what pool to use for swapping. Regardless if it is Pangolin or TraderJoeV2, we probably have to set it before we call the migrate() function.

I am also not worried about the gas cost of using oracle, but
1. the formula (and the code) is simplier when we use the Pangolin price
2. if we use the Pangolin price, then there will be a smaller 'dust' amount after addLiquidity
3. also have smaller gas cost which is not bad

the formula becomes this:
(ReserveToken2*AmountInToken1 - ReserveToken1*AmountInToken2)/(2*ReserveToken2)
for the case where we need to swap some Token1 to Token2

If you have difficulties deriving the formula (don't think that is the case), then I can write it down in a more readable way.

Reply:
If I were you, if I ever met anyone that had difficulty reading my handwriting in my life ever, then I would have written it. When was your last time you read an article written in handwriting in the internet? Let's not add unnecessary communication cost.

Your formula is the same with current one. The only difference is that you can replace Token2PriceInToken1 with ReserveToken1 and ReserveToken2 with the following definition `Token2PriceInToken1 = ReserveToken2 / ReserveToken1` in your context.

Actually if you use Pangolin, then you can remove an assumption that `amountOutToken2 = Token2PriceInToken1 * amountInToken1`. The right replacement is `amountOutToken2 = (ReserveToken2 * amountInToken1) / (ReserveToken1 + amountInToken1)`. And this makes code lengthier.

So I agree with your 3 but I don't agree with your 1 and 2.

I checked that there is wAVAX-USDC pool on TraderJoe V2. I'll ask Jasa about this and leave his answer here.

Reply:
I don't think this is the correct one unless there is 0% swap fee
amountOutToken2 = (ReserveToken2 * amountInToken1) / (ReserveToken1 + amountInToken1)

Using price oracle instead of the price implied by the pool would not be more accurate at all. The pricle has the potential to be way off compared to the price implied by the pool (both before and after the swap). 

We spent too much time already discussing this, if you want you can leave it as it is.

Reply:
Yeah I agree with you all though we can't know what is the oracle price after swap before we do swap. I think this discussion is going to split hairs, as we didn't make any meaningful step from the beginning, and I think that is because current solution is good enough. I'll leave this as is.